### Name: Lewis Vitta
### Username: LXV842
### ID number: 2047942
**please edit the above with your details to personlise this notebook**

# Laboratory A: Python Foundation: Variables, Keywords and Data Types

# Python for Engineers
(c) 2018,2019,2020 Dr Neil Cooke, School of Engineering, Collaborative Teaching Laboratory, University of Birmingham

# Introduction

This is foundation work. 

It consists of 8 short exercises that should take no more than 30-60 minutes to complete.

It is advised that you complete these before beginning the lab sessions. 

It introduces basic programming concepts and Python syntax.

Prior to this lab you should have:

* Begun the series of python lectures/videos
* Reviewed the online support materials for Python provided on Canvas
* Installed python on your personal computer and/or be familiar with its use.
* Be familiar with other procedural programming languages i.e. C.

You should upload this completed notebook to canvas. 

In the labs you will write a series of functions, which together, will simulate an electronic circuits.


## Learning Outcomes

This is a gentle introduction to the Python language and its syntax assuming that you have previous programming knowledge. You are advised to complete this before the lab. I cover the following:

> Variable names can only contain letters, number and underscores. Names cannot be started with a number.

> Variable types do not need to be stated explicitly in Python. The interpreter will assume a type depending on how the variable is declared.

> Python has 'reserved words' (or 'keywords') which cannot be used for the names of variables and constants.

> There are 5 built-in data types: Numbers, Strings, Lists, Tuples and Dictionaries

> Mathematical operators do not only have to be used with numbers, but can be used with other data types too (like strings)

> Python uses float division by default on integers, with floor division using a separate operator

> Data types can be changed in a program using casting

**tasks you are required to do are highlighted in bold throughout this notebook**

## Tools to complete this lab

* Use this Jupyter notebook and complete/save your work in it. This requires a computing platform capable of running Python 3.x and Jupyter notebook (Anaconda distribution is preferred at the time of writing). 

* NOTE: pressing CTRL+Z on a highlighted cell will return it to its previous state. Use this to undo any changes you have made if you get lost and want to start again.

* If for whatever reason you do not or cannot use Jupyter notebooks, then use the PDF document of this as a 'lab sheet' and type / copy the python code into a py file in the order presented in the sheet, using the IDLE python editor and suitable comments between sections e.g. #Exercise 1. This will require a computing platform with Python 3.x installed. (not recommended)


# Exercise 1: Variable Names

Some conventions:
* Variable names cannot be any series of characters and symbols. 
* There are a few rules which decide whether a variable name is legal. 
* Some variable names are legal, but defy convention. For example, 'ThisVariable' is a legal name for a variable, but using capital letters to separate the words (camel case) is not convention. 
* Variable and function names should be written with lower case letters, with words separated by underscores (snake case). 

**Try running the cells below (press shift enter when editing a cell to run it) to find which variable names are legal,and which are not. Change the variable names so they match the Python naming convention. **

In [1]:
power_10 = 4

In [2]:
voltage = 5.34 # V

In [3]:
total_current = 13.12 # A

In [4]:
no_of_loops = 8

In [5]:
reactive_power_ = 67.2 # W

In [6]:
impedance_start = 23.4 # Ohms

**the following function converts camel case to snake case. Run this function to load it in memory (shift-enter while in the cell below).**

In [2]:
import re # import a regular expression library with is used to parse text

#convert function takes in a CamelCase variable name and converts it to snake_case variable name
def convert(name):
    s1 = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)
    return re.sub('([a-z0-9])([A-Z])', r'\1_\2', s1).lower()

**now using the variable names you corrected previously, call the convert function with each one to determine new output.**

In [3]:
print(convert("No_Of_Loops")) #we are printing as multiple calls to functions in a cell will only result in the final call output
print(convert("KINDA FORGOT VARIABLE NAMES"))
print(convert("Doesn't add underscores ^")) 

no__of__loops
kinda forgot variable names
doesn't add underscores ^


# Exercise 2. Declaring Constants

* Constants should be declared at the top of a program. 
* The convention for constants is to use all caps with underscores to separate words. 
* However constants are just a variable the user doesnt change, and can be written using any legal variable name. 
* Unlike other languages (such as C), the data type of a variable does *not* need to be explicitly declared.

**Run the cells below and tweak the variable declarations to make the programs work correctly. 
Only the declarations need to be changed.**

In [9]:
DEFAULT_STRING = "Hello World" # initialised string prior to user input

string_entered = False # initialised flag to determine whether user has entered string

'''
in the real world there will be code here which does something- get user input 
'''

if string_entered == False: #user has not entered a string
    print(DEFAULT_STRING) # output the initialised string
    
#note: use flags in real programming only as a last resort!

Hello World


In [10]:
PI = 3.14 #value of PI to 2 dp

radius = 3 #radius of a circle
area_of_a_circle = PI * radius**2 #calculate areas of circle from radius using well known formula...

print(area_of_a_circle) # output the result.

28.26


# Exercise 3: Reserved Words

* Python has 'reserved words' (or 'key words') which have a built-in function. 
* Variables and constants cannot have the same name as these words or else the interpreter will be confused (hence the name 'reserved').

**The program below uses two variables with names which are the same as reserved words. Change the names of these variables so that the program can be run.**

In [11]:
R_wire = 0.6 #resistance of wire
L_wire = 0.3 #inductance of wire

#compute current in RL circuiut given real and imaginary voltages
def get_current(R, L, voltage_real, voltage_imag):
  R_equiv = R/3 
  L_equiv = L/3
  current_real = round(voltage_real/(R_equiv+R_wire),2)
  current_imag = round(voltage_imag - (L_equiv+L_wire),2)
  raise_voltage = "For a voltage of {} ∠ {} V : ".format(voltage_real,voltage_imag)
  return_current = "the current is {} ∠ {} A".format(current_real,current_imag)
  print(raise_voltage+return_current)

get_current(2000, 0.001, 5, 30)

For a voltage of 5 ∠ 30 V : the current is 0.01 ∠ 29.7 A


## void and fruitful functions

If you have used other programming languages, you may have noticed that this function does not have a return value; the function does not return anything when called. 

This is called a 'void function'. If a function does have a return value, it is called a 'fruitful function'.

These concepts will be covered more in depth in Lab 1.

# Exercise 4: Data Types

Python has 5 built in data types which you need to become very familiar with:

<center>*Numbers* - int, long, float, complex </center>
<center>*Strings* - created using either single quotes ('') or double quotes ("")</center>
<center>*Lists* - can store all data types, used to create arrays of numbers and text</center>
<center>*Tuples* - stores an ordered series of data which is immutable (cannot be changed)</center>
<center>*Dictionaries* - stores pairs of data called 'keys' and 'values', where calling a key will return its associated value</center>

* As mentioned in exercise 2, Python will assume the type of a variable based on how you declare it. 
* If it is delcared with inverted commas, it is assumed to be a string. 
* If it is declared with a decimal point, it is assumed to be a float. 

To check which type Python has assumed a variable (or object) to be, you can use the ```type()``` function call. 
* The argument this function takes is the name of the object you are testing. Using the ```type()``` function may come in useful when debugging a program.

**For an example, try running the cell below to determine its type (float)**

In [12]:
my_num = 1.5
type(my_num)

float

**Now use the ```type()``` function call to find the types of the variables below:**

In [13]:
circuit_arrangement = 'parallel'
type(circuit_arrangement)

str

In [14]:
initial_voltage = 5
type(initial_voltage)

int

In [15]:
sensor_output = [1.4,5.2,3.4,1.0,6.3,2.4]
type(sensor_output)

list

In [16]:
resistors = (180,1000,2500,3900)
type(resistors)

tuple

In [17]:
conversion_values = {'1':'0001','2':'0010','3':'0011','4':'0100'}
type(conversion_values)

dict

# Exercise 5: Basic Operations

* Data types can be operated on in different ways. 
* Each cell below uses a different operator to act on a variable(s). 

**See which data types can be operated on by which operator by adding and removing comments (#). Any line preceeded by a # will be ignored by the interpreter.**

In [18]:
''' Addition '''

x = 5
y = 6

x = 4.1
y = 2.8

x = 'Hello '
y = 'World!'

x = [2,1,3]
y = [4,2,4]
 
#x = (2,1,3)
#y = (4,2,4)

#x = {2:3,4:6,7:3}
#y = {3:1,5:3,1:0}



print(x+y)

[2, 1, 3, 4, 2, 4]


In [19]:
# Write a summary of what you found. For which data types is this operation valid? Which aren't?
#
#For the + operation all data types except the dictionary are valid. 
#
#
#
#
#
#
#


In [20]:
''' Subtraction '''

x = 3
y = 9

x = 2.5
y = 6.7

##x = 'Subtract '
##y = 'Strings'

##x = [5,3,6]
##y = [2,7,4]
 
##x = (5,3,6)
##y = (2,7,4)

#x = {'one':'uno','two':'dos','three':'tres'}
#y = {'four':'cuatro','five':'cinco','six':'seis'}



print(x-y)

-4.2


In [21]:
# Write a summary of what you found. For which data types is this operation valid? Which aren't?
#
# the - operator is only valid for Ints and Floats
#
#
#
#
#
#
#


In [22]:
''' Multiplication '''

x = 6

x = 3.4

x = 'Multiplication'

x = [3,4,7]

x = (3,4,7)

#x = {2:4,5:7,8:10}



print(x*4)

(3, 4, 7, 3, 4, 7, 3, 4, 7, 3, 4, 7)


In [23]:
# Write a summary of what you found. For which data types is this operation valid? Which aren't?
#
#
#
#All types are valid for integer multiplication, except the dictionary 
#
#
#
#
#


In [24]:
''' Division '''

x = 9

x = 6.3

#x = 'Division'

#x = [5,7,3]

#x = (5,7,3)

#x = {4:3,6:2,8:6}



print(x/4)

1.575


**Write a summary of what you found. For which data types is this operation valid? Which aren't?**

 For division integers and floats are both valid. for 9/4 or any non-integer answer python will automatically convert to a float (float division)




In [25]:
''' Floor Division '''
# This operation only works with numbers. Some languages, like C, will treat integers differently when they are divided.
# The default type of division is 'integer division' (or 'floor division'). Python however, will always perform floating
# point division by default, unless the // operator is used. Change the operators to see Python handles division for
# integers and floats.


x = 72.5
y = 12

print(x/y)


a = 6
b = 4

print(a/b)

6.041666666666667
1.5


**Write a summary of what you found. For which data types is this operation valid? Which aren't?**
for floor division "//" flaots and ints are both valid. Floor division will round down to the integer value from whatever the float division answer is.


In [26]:
''' Modulo '''
# Again, modulo only works with numbers. This cell uses the same numbers as the previous.
# Find the difference between the // and % operators. Is there any similarity between the two?

x = 3.5
y = 4

print(x%y)


a = 21
b = 4

print(a/b)

3.5
5.25


**Write a summary of what you found. For which data types is this operation valid? Which aren't?**
Modulo will return the remainder from division, similarly floor division is the whole number from division.

take a = 21, b = 4. the floor returns 5, the modulo 1. 21/4 is 5 remainder 1. or 5.25

# Exercise 6: Assignment and Equality Operators

* Python uses two operators which are very similar, but have very different meanings. 
* These are the '=' and '==' operators, or the assignment and equality operators respectively.

*The assignment operator should only be used when setting a variable equal to a value. This must be done from left to right, unlike mathematics. For example, writing 'x = 4' will set the variable equal to four. But writing '4 = x' (right to left assignment) is not valid and will return an error.

*The equality operator checks whether two values are the same. The values can be stored in variables, or just be the plain value. For example, you can type 'x == 4' and 'x == y', both of these are valid.

**In the cell below, check if you can use right to left comparison of variables, that is, you can write the value first, then the variable.**

In [27]:
x = 4
if 4==x:
    print("Right to left comparison is valid")
else:
    print("Right to left comparison is invalid")

Right to left comparison is valid


# Exercise 7: Casting


* Sometimes when a variable is defined it is of one type, but later in the program the type needs to change, either for an operation to work or to make a comparison etc. 
* Converting a data type from one to another is called 'casting'.
* To convert to a float, integer or string, use the functions float(), int() and str() respectively. 
* The argument for this function is simply the object you want to cast.

**Using the three casting functions listed above, change the data types of the variables in the programs below to remove the type errors.**

In [28]:
part_number = 1005
part_name = 'BSKDAM'


product_code = part_name + str(part_number)

print(product_code)

BSKDAM1005


In [29]:
decimal_number = '45.215523'

remainder = float(decimal_number)%1

print(remainder)

0.21552299999999747


In [30]:
user_age = '24'

if int(user_age) >= 11:
    print("You can make an account")
else:
    print("You cannot make an account")

You can make an account


# Exercise 8: User Input

* Values can be assigned to variables by obtaining input from a user, for example through using a keyboard. 
* Python uses the function 'input()' where the argument is text to be displayed on screen, e.g "Enter Value: ". All data stored using the input() command is stored as a string. 
* In order to use input for calculations, if the user inputs a number, the data must be cast.

**Find and fix the errors in the program below. In the remaining cell, write a program which takes in a number from the user, and prints the value of the number squared.**

In [36]:
num_of_stars = input("How many stars would you like? : ") #get number of stars from user
star_string = '' #initialise variable star string

for x in range(int(num_of_stars)): #for every star
    star_string += '*' #add a star to the star string
    
print(star_string) #output the star string to the user

How many stars would you like? : 5
*****


**Write a program which asks the user to enter a number, then prints out the value of the number squared:**

In [38]:
# your program here

x = input("Enter a number ")
print(float(x)**2)






Enter a number 69.420
4819.1364


# End of Lab


This is the end of your introduction to Variables, Keywords and Data Types. 

Please note this is not a complete reference but is designed to introduce you to concepts and common misconceptions of the python language, and hopefully some good practice such as variable naming and exhaustive commenting of your code...

Please consult the online reference materials to find out more (Canvas has links).

Jupyter notebooks are not just for coding; you can write notes like these in markup language. 

So note in the next cell what what you learned/found difficult or easy in the space below.


**please reflect on your learning below**

### I rate this lab (out of 5): 5

### What I find easy:
All of the above



# Next steps

* **Upload this completed .pynb notebook to canvas**
* **Progress to the next lab**


# Python for Engineers
(c) 2018, 2019, 2020 Dr Neil Cooke, School of Engineering, Collaborative Teaching Laboratory, University of Birmingham